Fix leaking goroutines in Install

mslapek opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/main/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->


**What this PR does / why we need it**:

Closes  #11971

**Special notes for your reviewer**:

No public API was changed.

Earlier `failRelease` could be called twice (when `ctx` was cancelled and the background job had an error).
Now `failRelease` can be called at most once - synchronized with a channel (instead of a mutex lock).

Added comment with true behavior - running the job in the background after cancel is a bit surprising...

Done small refactoring. All fan-in logic is in a small function.

**If applicable**:
- [ ] this PR contains documentation
- [X] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

gjenkins8 reviewed (commented):
To me (community member; not a helm maintainer), this looks like a fairly nice change generally. It refactors and improves significant amount of messy go-routine and error handling/synchronization logic. Plus looks to fix the mentioned go-routine leak (bug).

I do think this PR somewhat exposes a larger issue, cancelling the context doesn't really 'cancel' the install (or upgrade). It does "fail" the install/upgrade  (calls the respective `failRelease`), but perhaps worse, likely the install/upgrade is failed whilst the install/update is still (likely) running. As my comment, I think this is a seperate issue from this PR.

I will raise both this PR, and the async/cancel issue in the helm weekly developer call.

mslapek commented on pkg/action/install.go:
```diff
@@ -211,6 +211,9 @@ func (i *Install) Run(chrt *chart.Chart, vals map[string]interface{}) (*release.
 
 // Run executes the installation with Context
+//
+// When the task is cancelled through ctx, the function returns and the install
```
> Completely agree, this is surprising.

I was myself surprised with that - so added a comment for other devs. 😉

> And I think worth discussing further with helm maintainers.

Looks like the discussion started in issue https://github.com/helm/helm/issues/12086

hiddeco replied:
:scream: we missed this as our whole container gets killed in scenarios where a context is cancelled, but I would definitely like to see this addressed as well.

mattfarina commented on pkg/action/install.go:
```diff
@@ -368,63 +371,76 @@ func (i *Install) RunWithContext(ctx context.Context, chrt *chart.Chart, vals ma
+	doneChan := make(chan struct{})
+	defer close(doneChan)
+
+	go func() {
+		rel, err := i.performInstall(rel, toBeAdopted, resources)
+		resultChan <- Msg{rel, err}
+	}()
+	go func() {
+		select {
+		case <-ctx.Done():
+			err := ctx.Err()
+			resultChan <- Msg{rel, err}
+		case <-doneChan:
+			return
+		}
+	}()
```
When the context is canceled the second goroutine will exit but it looks like the first will still run until `performInstall` reaches completion. Am I missing something?

I wonder if `performInstall` needs to be modified to accept a context and regularly check if it's done to exit.

mslapek replied:
> When the context is canceled the second goroutine will exit but it looks like the first will still run until performInstall reaches completion.

This is right - it was the original behavior even before this PR. This was captured by the added comment to the function. I think it's a technical debt.

> I wonder if performInstall needs to be modified to accept a context and regularly check if it's done to exit.

Yes. The API communicating with K8s (like `KubeClient.WaitWithJobs`) would need to take context instead of timeout, some people consider this in https://github.com/helm/helm/issues/12086

I think that this PR might be a good starting point to fix the mentioned issue (separetes the code into `releasingUpgradeCtx` and `performInstallCtx`).

gjenkins8 replied:
I think this PR (as is) fixes lost goroutines in performInstall/Upgrade.

https://github.com/helm/helm/issues/12086 (and associate PR: https://github.com/helm/helm/pull/12109) might be a better place to tackle synchronizing install.upgrades to not continue to run in the background?

gjenkins8 replied:
I think this PR (as is) fixes lost goroutines in performInstall/Upgrade.

https://github.com/helm/helm/issues/12086 (and associate PR: https://github.com/helm/helm/pull/12109) might be a better place to tackle synchronizing install.upgrades to not continue to run in the background?

scottrigby commented:
I pinged @hiddeco to see if he can review this as a heavy Helm SDK user. I recall goroutines and client workarounds were needed in Flux's use of Helm SDK.

hiddeco commented on pkg/action/install.go:
```diff
@@ -368,63 +371,76 @@ func (i *Install) RunWithContext(ctx context.Context, chrt *chart.Chart, vals ma
+		rel, err := i.performInstall(rel, toBeAdopted, resources)
+		resultChan <- Msg{rel, err}
+	}()
+	go func() {
+		select {
+		case <-ctx.Done():
+			err := ctx.Err()
+			resultChan <- Msg{rel, err}
+		case <-doneChan:
+			return
+		}
+	}()
```
Think this whole bit (and below) is an obsolete routine? AFAIK, the following should work without any issues while yielding the same behavior and getting rid of a `chan`:

```go
select {
case <-ctx.Done():
	return rel, ctx.Err()
case msg := <-resultChan:
	return msg.r, msg.e
}

joejulian commented:
Is there any way to add regression tests for this?

tsuna commented:
Is this holding up the 3.12.2 release?

mslapek commented:
## Recent changes

* limited the scope of this PR to Install action only (was Install and Upgrade previously)
* added unit-tests checking for leaking goroutines @joejulian
* applied a hint from @hiddeco (thanks!)

## Tests

Added to a few tests inspection of [`runtime.NumGoroutine()`](https://pkg.go.dev/runtime#NumGoroutine).

They nicely fail on main:

```
$ go clean -testcache && go test ./pkg/action

--- FAIL: TestInstallRelease_Wait (0.00s)
    install_test.go:380: 
                Error Trace:    /home/awesome/repo/helm/pkg/action/install_test.go:380
                Error:          Not equal: 
                                expected: 2
                                actual  : 3
                Test:           TestInstallRelease_Wait
--- FAIL: TestInstallRelease_Wait_Interrupted (11.01s)
    install_test.go:405: 
                Error Trace:    /home/awesome/repo/helm/pkg/action/install_test.go:405
                Error:          Not equal: 
                                expected: 2
                                actual  : 3
                Test:           TestInstallRelease_Wait_Interrupted
```

(the tests show one additional running goroutine - the leaking one!)

Of course, they should succeed on this branch. 😉

joejulian reviewed (approved):
Nice! Thanks

mslapek committed:
Fix leaking goroutines in Install

Signed-off-by: Michał Słapek <28485371+mslapek@users.noreply.github.com>

mattfarina commented:
@mslapek can you rebase this. When I went to merge it there was a conflict.

mslapek commented:
> @mslapek can you rebase this. When I went to merge it there was a conflict.

Rebased. The conflict was a trivial new-line clash of added isDryRun from main.

mslapek commented:
@joejulian @mattfarina @hiddeco Thanks for review! 🍭
